### PR TITLE
Redirects: allow to redirect even if a page exists

### PR DESCRIPTION
**Pull request by stsewd**
Netify call this option "force", I think it is okay to call them that way in our platform.

- This is currently under a feature flag (should I remove the docs till we make this feature available to everyone? or maybe just mention that's a beta feature)
- Redirects are cached as normal as the other requests (there is another PR on rtd-ext to purge the cache)

### CODE REVIEW

File: readthedocs/projects/forms.py [hunk 1]
```diff
@@ -585,21 +585,30 @@ class RedirectForm(forms.ModelForm):
...
 
     def __init__(self, *args, **kwargs):
         self.project = kwargs.pop('project', None)
         super().__init__(*args, **kwargs)
 
+        if self.project.has_feature(Feature.ALLOW_FORCED_REDIRECTS):
+            # Remove the nullable option from the form.
+            # TODO: remove after migration.
+            self.fields["force"].widget = forms.CheckboxInput()
+            self.fields["force"].empty_value = False
```
Comment (stsewd) [inline readthedocs/projects/forms.py:598]: We could also make the field non-nullable and just deal with a couple of minutes of users no being able to create redirects.

File: readthedocs/proxito/views/serve.py [hunk 2]
```diff
@@ -176,16 +176,25 @@ def get(self,
...
-        # redirect_path, http_status = self.get_redirect(
-        #     final_project,
-        #     lang_slug,
-        #     version_slug,
-        #     filename,
-        #     request.path,
-        # )
-        # if redirect_path and http_status:
-        #     return self.get_redirect_response(request, redirect_path, http_status)
+        redirect_path, http_status = self.get_redirect(
```
Comment (stsewd) [inline readthedocs/proxito/views/serve.py:179]: We could just call `final_project.get_redirect_path_with_status()` directly, we aren't adding anything extra on `get_redirect`.

File: readthedocs/proxito/views/serve.py [hunk 3]
```diff
@@ -304,26 +313,6 @@ def get(self, request, proxito_path, template_name='404.html'):
...
-        # For example if the filename is ``http://example.com/my-path``,
-        # ``.path`` would return ``my-path``.
-        parsed = urlparse(filename)
-        redirect_filename = parsed._replace(query="").geturl()
-
-        # We can't check for lang and version here to decide if we need to add
-        # the ``/`` or not because ``/install.html`` is a valid path to use as
-        # redirect and does not include lang and version on it. It should be
-        # fine always adding the ``/`` to the beginning.
-        redirect_filename = '/' + redirect_filename.lstrip('/')
```
Comment (stsewd) [inline readthedocs/proxito/views/serve.py:325]: this was moved to the get_redirect_path_with_status method itself

File: readthedocs/proxito/views/serve.py [hunk 4]
```diff
@@ -389,15 +378,15 @@ def get(self, request, proxito_path, template_name='404.html'):
...
+                        path=filename,
                         full_path=proxito_path,
                     )
                     return resp
 
         self._register_broken_link(
             project=final_project,
             version=version,
-            path=redirect_filename,
+            path=filename,
```
Comment (stsewd) [inline readthedocs/proxito/views/serve.py:389]: We already normalize the filename before recording the broken link, so we are fine using the original filename

File: docs/user/user-defined-redirects.rst [hunk 5]
```diff
@@ -128,6 +130,17 @@ Similarly, if you maintain several branches of your documentation (e.g. ``3.0``
 ``latest``) and decide to move pages in ``latest`` but not the older branches, you can use
 *Exact Redirects* to do so.
 
+You can use an exact redirect to migrate your documentation to another domain,
```
Comment (ericholscher) [inline docs/user/user-defined-redirects.rst:133]: This section needs some subsections. There's a lot of content here, I think a few subsections would make it easier to follow. Probably headers with what each example is doing?

File: readthedocs/projects/forms.py (hunk 1 above)
Comment (ericholscher) [reply readthedocs/projects/forms.py:598]: No strong preference here, but being safe seems fine.

File: readthedocs/proxito/views/serve.py [hunk 6]
```diff
@@ -176,16 +176,25 @@ def get(self,
...
+        if redirect_path and http_status:
+            try:
+                return self.get_redirect_response(
+                    request=request,
+                    redirect_path=redirect_path,
+                    proxito_path=request.path,
+                    http_status=http_status,
+                )
+            except InfiniteRedirectException:
+                # Continue with our normal serve.
```
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:196]: Should we log something here?

File: readthedocs/redirects/querysets.py [hunk 7]
```diff
@@ -35,16 +36,52 @@ def api(self, user=None):
...
+        :param path: Is the path without the language and version parts.
+        :param full_path: Is the full path including the language and version parts.
+        :param forced_only: Include only forced redirects in the results.
+        """
+
+        # Small optimization to skip executing the big query below.
+        if forced_only and not self.filter(force=True).exists():
+            return None, None
+
+        def normalize_path(path):
```
Comment (ericholscher) [inline readthedocs/redirects/querysets.py:54]: This inline function feels a bit weird.. Is there a reason it isn't in a utils file?

File: docs/user/user-defined-redirects.rst [hunk 8]
```diff
@@ -3,28 +3,30 @@ User-defined Redirects
...
-If you need to redirect a large number of files that still exist,
-please reach out to :doc:`/support`.
+Features
+--------
 
-Page & Exact Redirects can redirect to URLs outside Read the Docs.
-Define the `To URL` as the absolute URL you want to redirect to.
+- By default, redirects are followed only if the requested page doesn't exist
+  (*404 File Not Found* error), if you need to apply a redirect for files that exist,
+  mark the :guilabel:`Force redirect` option.
```
Comment (ericholscher) [inline docs/user/user-defined-redirects.rst:26]: A nicer way to talk about something behind a feature flag :D 

```suggestion
  mark the :guilabel:`Force redirect` option. **This option is only available on some plan levels**.
  Please ask support if you need it for some reason. 
```

**Review by ericholscher — APPROVED** [review]
This looks straightforward. I'd like to get it deployed so we can test it out for performance. I think it's probably going to be painful, but it's useful to have if we need it for a user, for now.

File: readthedocs/redirects/querysets.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/redirects/querysets.py:54]: it was very specific to this method, I'll move it to be a private method

File: readthedocs/proxito/views/serve.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:196]: We already log that here https://github.com/readthedocs/readthedocs.org/blob/0acd5e4cfe2369792ba03338f81aa81c23bd5c66/readthedocs/proxito/views/mixins.py#L248-L254

but we can add another here to be more specific

File: docs/user/user-defined-redirects.rst (hunk 5 above)
Comment (stsewd) [reply docs/user/user-defined-redirects.rst:150]: I have put the examples in subsections.

Screenshot 2022-05-26 at 16-53-47 User-defined Redirects — Read the Docs user documentation 8 0 2 documentation

There is a lot of nesting, maybe we could address a more structural change as part of https://github.com/readthedocs/readthedocs.org/issues/9258

File: readthedocs/proxito/views/serve.py (hunk 6 above)
Comment (humitos) [reply readthedocs/proxito/views/serve.py:196]: We should add `force=` attribute to that log line to have more context when debugging.

**Pull request merged**